File: common/src/main/java/org/sandboxpowered/silica/block/EnumProperty.java
package org.sandboxpowered.silica.block;

import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Lists;
import org.sandboxpowered.api.state.property.AbstractProperty;

import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class EnumProperty<E extends Enum<E>> extends AbstractProperty<E> {
    private final ImmutableSet<E> values;
    private final Map<String, E> names = new HashMap<>();

    private EnumProperty(String name, Class<E> type, Collection<E> collection) {
        super(name, type);
        this.values = ImmutableSet.copyOf(collection);
        values.forEach(e -> {
            names.put(getName(e), e);
        });
    }

    public static <T extends Enum<T>> EnumProperty<T> of(String name, Class<T> type) {
        return of(name, type, t -> true);
    }

    public static <T extends Enum<T>> EnumProperty<T> of(String name, Class<T> type, Predicate<T> predicate) {
        return of(name, type, Arrays.stream(type.getEnumConstants()).filter(predicate).collect(Collectors.toList()));
    }

    public static <T extends Enum<T>> EnumProperty<T> of(String name, Class<T> type, T... values) {
        return of(name, type, Lists.newArrayList(values));
    }

    public static <T extends Enum<T>> EnumProperty<T> of(String name, Class<T> type, Collection<T> values) {
        return new EnumProperty<>(name, type, values);
    }

    @Override
    public String getName(E value) {
        return value.name().toLowerCase(Locale.ENGLISH); //TODO: replace with proper name methods
    }

    @Override
    public Collection<E> getValues() {
        return values;
    }

    @Override
    public Optional<E> getValue(String name) {
        return Optional.ofNullable(names.get(name));
    }
}

File: common/src/main/kotlin/org/sandboxpowered/silica/state/property/EnumProperty.kt
package org.sandboxpowered.silica.state.property

import com.google.common.collect.ImmutableSet
import com.google.common.collect.Lists
import java.util.*
import java.util.function.Consumer
import java.util.function.Predicate
import java.util.stream.Collectors

class EnumProperty<E> private constructor(name: String, type: Class<E>, collection: Collection<E>) :
    AbstractProperty<E>(name, type) where E : Enum<E>, E : StringSerializable {
    override val values: ImmutableSet<E>
    private val names: MutableMap<String, E> = HashMap()
    override fun getValueString(value: E): String = value.getName()

    override fun getValue(name: String): Optional<E> = Optional.ofNullable(names[name])

    companion object {
        @JvmStatic
        fun <T> of(name: String, type: Class<T>): EnumProperty<T> where T : Enum<T>, T : StringSerializable {
            return of(name, type) { true }
        }

        @JvmStatic
        fun <T> of(
            name: String,
            type: Class<T>,
            predicate: Predicate<T>?
        ): EnumProperty<T> where T : Enum<T>, T : StringSerializable {
            return of(name, type, Arrays.stream(type.enumConstants).filter(predicate).collect(Collectors.toList()))
        }

        @JvmStatic
        fun <T> of(
            name: String,
            type: Class<T>,
            vararg values: T
        ): EnumProperty<T> where T : Enum<T>, T : StringSerializable {
            return of(name, type, Lists.newArrayList(*values))
        }

        @JvmStatic
        fun <T> of(
            name: String,
            type: Class<T>,
            values: Collection<T>
        ): EnumProperty<T> where T : Enum<T>, T : StringSerializable {
            return EnumProperty(name, type, values)
        }
    }

    init {
        values = ImmutableSet.copyOf(collection)
        values.forEach(Consumer { e: E -> names[getValueString(e)] = e })
    }
}